### ISSUE TITLE
`reportPossiblyUnboundVariable` false negative in inner function that refers to nonlocal variable if the function is defined before the variable is assigned

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#773]
Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=CYUwZgBAhgFAlALgLACgLomA9gJwgNygBsBXECASwDsIiKBnAFwG1rGBdeZNDXgBxxsYhUiDipUvUJGpUQOLpN4YBQkWTgQAxBHk5cCCDhB9cjAApZ69CgCMiATwCqVW1hJVgANSiCo9kAkedFl5eCDpCFtFYIhI0IVEJWUIVSpGYWINZMxcAizyaloGFjZOJNj%2BQXTM0U0dKixdHH0cIN4E%2BCA)

```python
def a():
    for value in list[int]():
        print(value)

    def inner():
        print(value) # error: reportPossiblyUnboundVariable

    inner()

def b():
    def inner():
        print(value)
    for value in list[int]():
        print(value) # no error

    inner()
```
